File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> temp;
        combos( candidates , res , temp , 0 , target);
        return res;
    }
    
    void combos(vector<int>& candidates, vector<vector<int>>& res , vector<int> temp , int x , int target){
        if( target == 0){
            res.push_back(temp);
        }
        
        for(int i = x ; i < candidates.size() ; i++){
            if( target >= candidates[i]){
                temp.push_back(candidates[i]);
                combos( candidates , res , temp , i , target - candidates[i]);
                temp.pop_back();
            }
        }
    }
};

File: 0039-combination-sum/0039-combination-sum.go
func combinationSum(candidates []int, target int) [][]int {
    sort.Ints(candidates) // Sort candidates to enable pruning

    var res [][]int
    var path []int

    var dfs func(start int, target int)
    dfs = func(start int, target int) {
        if target == 0 {
            // Found valid combination
            combination := make([]int, len(path))
            copy(combination, path)
            res = append(res, combination)
            return
        }

        for i := start; i < len(candidates); i++ {
            candidate := candidates[i]
            
            // Prune: if candidate > target → no point in continuing
            if candidate > target {
                break
            }

            path = append(path, candidate)
            dfs(i, target - candidate) // Reuse the same element
            path = path[:len(path)-1]   // Backtrack
        }
    }

    dfs(0, target)
    return res
}
